# Free GPUs for All

The purpose is to showcase the power and flexibility of Google Colab to solve a classification problem. We will be leveraging Tensorflow's Imagenet pre-trained model to complete our task.

### Setup

Importing Tensorflow

In [0]:
import tensorflow as tf

Checking version of Tensorflow

In [3]:
pip list | grep tensorflow

DEPRECATION: The default format will switch to columns in the future. You can use --format=(legacy|columns) (or define a format=(legacy|columns) in your pip.conf under the [list] section) to disable this warning.
tensorflow (1.4.1)
tensorflow-tensorboard (0.4.0)


Ensuring GPU setup for this notebook

In [2]:
tf.test.gpu_device_name()

'/device:GPU:0'

### Download Pre-Trained Model

In [0]:
git clone https://github.com/tensorflow/models.git

Cloning into 'models'...
remote: Counting objects: 11513, done.
remote: Compressing objects: 100% (7/7), done.
remote: Total 11513 (delta 2), reused 2 (delta 2), pack-reused 11504
Receiving objects: 100% (11513/11513), 313.91 MiB | 42.61 MiB/s, done.
Resolving deltas: 100% (6229/6229), done.


### Running Our Model

Directory where the imagenet model is located:

In [0]:
cd models/models/tutorials/image/imagenet/

/models/models/tutorials/image/imagenet


Saving a picture of a basketball to our instance. We will use this image with the pre-trained imagenet classifier.

In [0]:
wget http://3.bp.blogspot.com/_qjlkTyHLyuw/TFt3adewwrI/AAAAAAAAAAk/Bdz2jz79Yjs/s1600/Basketball-large.png

--2018-02-07 22:47:09--  http://3.bp.blogspot.com/_qjlkTyHLyuw/TFt3adewwrI/AAAAAAAAAAk/Bdz2jz79Yjs/s1600/Basketball-large.png
Resolving 3.bp.blogspot.com (3.bp.blogspot.com)... 74.125.141.132, 2607:f8b0:400c:c06::84
Connecting to 3.bp.blogspot.com (3.bp.blogspot.com)|74.125.141.132|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 92358 (90K) [image/png]
Saving to: ‘Basketball-large.png’

Basketball-large.pn 100%[===================>]  90.19K  --.-KB/s    in 0s      

2018-02-07 22:47:09 (442 MB/s) - ‘Basketball-large.png’ saved [92358/92358]



Running pre-trained classification model on our image of a basketball

In [0]:
python classify_image.py --image_file Basketball-large.png

2018-02-07 22:47:25.031609: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:892] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2018-02-07 22:47:25.031962: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1030] Found device 0 with properties: 
name: Tesla K80 major: 3 minor: 7 memoryClockRate(GHz): 0.8235
pciBusID: 0000:00:04.0
totalMemory: 11.17GiB freeMemory: 505.38MiB
2018-02-07 22:47:25.032000: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1120] Creating TensorFlow device (/device:GPU:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:00:04.0, compute capability: 3.7)
basketball (score = 0.99916)
orange (score = 0.00011)
lifeboat (score = 0.00006)
monarch, monarch butterfly, milkweed butterfly, Danaus plexippus (score = 0.00006)
space heater (score = 0.00003)


### Results

Our model is 99.9% certain that our basketball image is in fact a basketball. Feel free to test the model with any image and see what you get. The imagenet model is pretrained with ~1,000 categories.